packages/controllers/src/snaps/SnapController.ts

```diff
@@ -726,6 +799,119 @@ export class SnapController extends BaseController<
     }, this._idleTimeCheckInterval) as unknown as number;
   }
 
+  /**
+   * Checks all installed and unblocked snaps against the block list and blocks
+   * snaps as appropriate. See {@link SnapController.blockSnap} for more
+   * information.
+   */
+  async updateBlockedSnaps(): Promise<void> {
+    if (!this._checkSnapBlockList) {
+      return;
+    }
+
+    const blockedSnaps = await this._checkSnapBlockList(
+      Object.values(this.state.snaps).reduce((blockListArg, snap) => {
+        if (!snap.blocked) {
+          blockListArg[snap.id] = snap.version;
+        }
+
+        return blockListArg;
+      }, {} as Record<SnapId, SemVerVersion>),
+    );
+
+    if (Object.keys(blockedSnaps).length === 0) {
+      return;
+    }
+
+    await Promise.all(
+      Object.entries(blockedSnaps).map(
+        ([snapId, { blocked, infoUrl, reason }]) => {
+          if (blocked) {
+            return this._blockSnap(snapId, { infoUrl, reason });
+          }
+
+          this._unblockSnap(snapId);
+          return Promise.resolve();
+        },
+      ),
+    );
+  }
+
+  /**
+   * Blocks an installed snap and prevents it from being started again. Emits
+   * {@link SnapBlocked}. Does nothing if the snap is not installed.
+   *
+   * @param snapId - The snap to block.
+   * @param blockedSnapInfo - Information detailing why the snap is blocked.
+   */
+  private async _blockSnap(
+    snapId: SnapId,
+    blockedSnapInfo: BlockedSnapInfo,
+  ): Promise<void> {
+    if (!this.has(snapId)) {
+      return;
+    }
+
+    try {
+      this.update((state: any) => {
+        state.snaps[snapId].blocked = true;
+        state.snaps[snapId].blockInformation = blockedSnapInfo;
+      });
+
+      await this.disableSnap(snapId);
```

Comment:
It may be confusing to disable/enable a snap when it is blocked/unblocked. Because if it is unblocked it may be enabled even though it was disabled originally. Is there a need to enable/disable when we can just fail if it is blocked?

Reply:
Here's my thinking about this:

I believe that we should continue to use enablement to determine whether a snap can be started, and I don't want to have two separate booleans that determine whether a snap can be started or not. We have a couple of dedicated methods for enabling/disabling a snap, and we can just add the `snap.blocked` check to determine whether a snap can be enabled, which is what I've done.

Now, if enablement/disablement ultimately determines whether a snap can be started, then blocking a snap must disable it. You are right to question whether a snap should become enabled when it is unblocked, however.

So far, I removed the `enableSnap()` call when an installed snap becomes unblocked during a blocklist check. The `:snapUnblocked` event is emitted when that happens, and we can prompt the user to re-enable a snap in that case, which in any case should be extremely rare. The idea here is that a user may have disabled a snap before it became blocked, or have forgotten about it while it was blocked etc., and automatically enabling it would be a violation of user consent. So, I removed that.

However, I did choose to enable the snap when a blocked snap is successfully updated to an unblocked version. The idea here is that, if you are updating the snap, you are trying to use it, and you are (at least currently) receiving the update request from a website. So, it makes sense that the snap is enabled and then started when it is updated. We can always revisit this and give the user the option to enable the snap, but it's hard to make a decision without recording if the snap was disabled manually or just when the snap became blocked.

What do you think?